File: libs/toolkit/src/test/java/com/example/toolkit/OptionalTest.java
package com.example.toolkit;

import org.hamcrest.MatcherAssert;
import org.hamcrest.core.Is;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import static org.hamcrest.MatcherAssert.assertThat;

@RunWith(JUnit4.class)
public class OptionalTest {
	@Test
	public void isPresentShouldReturnFalseWhenNullOptueIsProvided() {
		Optional maybe = Optional.fromNullable(null);

		assertThat(maybe.isPresent(), Is.is(false));
		assertThat(maybe.isNotPresent(), Is.is(true));
	}

	@Test
	public void isPresentShouldReturnTrueWhenOptidOptueIsProvided() {
		Optional maybe = Optional.from("test data");

		assertThat(maybe.isPresent(), Is.is(true));
		assertThat(maybe.isNotPresent(), Is.is(false));
	}

	@Test(expected = NullPointerException.class)
	public void shouldThrowExceptionWhenCallingFromWithNullOptue() {
		Optional.from(null);
	}

	@Test
	public void emptyShouldReturnInstanceWithoutData() {
		Optional maybe = Optional.empty();

		assertThat(maybe.isPresent(), Is.is(false));
		assertThat(maybe.isNotPresent(), Is.is(true));
	}

	@Test
	public void getShouldReturnProvidedData() {
		Optional<String> maybe = Optional.from("test");
		String data = maybe.get();

		MatcherAssert.assertThat(data, Is.is("test"));
	}

	@Test
	public void getOrDefaultShouldReturnTheDefaultOptueWhenMaybeIsEmpty() {
		Optional<String> maybe = Optional.fromNullable(null);
		String data = maybe.getOrDefault("test");

		MatcherAssert.assertThat(data, Is.is("test"));
	}

	@Test
	public void getOrDefaultShouldReturnProvidedDataWhenMaybeIsNotEmpty() {
		Optional<String> maybe = Optional.fromNullable("test");
		String data = maybe.getOrDefault("test 2");

		MatcherAssert.assertThat(data, Is.is("test"));
	}
}

File: libs/toolkit/src/test/kotlin/br/com/vitorsalgado/example/toolkit/OptionalTest.kt
package br.com.vitorsalgado.example.toolkit

import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.core.Is.`is`
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class OptionalTest {
  @Test
  fun `"isPresent" should return false when null source is provided`() {
    val maybe = Optional.fromNullable<Any>(null)

    assertThat(maybe.isPresent, `is`(false))
    assertThat(maybe.isNotPresent, `is`(true))
  }

  @Test
  fun `"isPresent" should return true when source is not null`() {
    val maybe = Optional.from("test data")

    assertThat(maybe.isPresent, `is`(true))
    assertThat(maybe.isNotPresent, `is`(false))
  }

  @Test(expected = NullPointerException::class)
  fun `it should throw exception when calling "from" with null source`() {
    Optional.from<Any>(null)
  }

  @Test
  fun `"empty" should return instance without data`() {
    val maybe = Optional.empty<Any>()

    assertThat(maybe.isPresent, `is`(false))
    assertThat(maybe.isNotPresent, `is`(true))
  }

  @Test
  fun `"get" should return source data`() {
    val maybe = Optional.from("test")
    val data = maybe.get()

    assertThat<String>(data, `is`("test"))
  }

  @Test
  fun `"getOrDefault" should return the default value when source is empty`() {
    val maybe = Optional.fromNullable<String>(null)
    val data = maybe.getOrDefault("test")

    assertThat<String>(data, `is`("test"))
  }

  @Test
  fun `"getOrDefault" should return source data when optional is not empty`() {
    val maybe = Optional.fromNullable("test")
    val data = maybe.getOrDefault("test 2")

    assertThat<String>(data, `is`("test"))
  }
}
